### PR TITLE
Fix problem with unknown characters when detecting line endings

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?
Fixes issue with the `render_tmpl` function on Windows. A portion of the code looks for a line ending at the end of a file. If there are any recognizable characters in the file it would crash. Since we only care about the end, we're going to use `repr`.

### What issues does this PR fix or reference?
Found in testing

### Tests written?
No

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by terminalmage — COMMENTED** [review]
I don't really feel qualified to review this since I don't have Windows available to test this.

**Comment by damon-atkins** [comment]
is their a function in windows to check if a file is "not open", rather than change the test, should this be a changed to the function being tested to retry.

**Comment by twangboy** [comment]
@damon-atkins It's not testing if the file is open, it's testing if the file ends with a newline character or not. If it finds a newline character at the end, then it makes sure to add a new line character when joining the file back together. The contents don't matter, just the ending.

**Comment by damon-atkins** [comment]
Sorry looks like I commented against the wrong PR

**Comment by rallytime** [comment]
re-run lint

**Comment by damon-atkins** [comment]
@terminalmage  I can not review this as its code I have not looked at before.
However Windows can have both  `\n` and `\r\n` files.   If a file is edited then line feed should not be changed.   If a file comes from the master (does not exist already on system) then then the line feed should not be changed.  i.e. on the master the person should have put `\r\n` in the file.

**Comment by twangboy** [comment]
I believe the python open command will replace line endings in a file with the system line endings.

File: salt/utils/templates.py [hunk 1]
```diff
@@ -207,7 +207,8 @@ def render_tmpl(tmplsrc,
                 output = output.encode(SLS_ENCODING)
             if salt.utils.platform.is_windows():
                 newline = False
-                if output.endswith(('\n', os.linesep)):
+                # Use repr as we don't care about the contents, only the ending
+                if repr(output).endswith(('\n', os.linesep)):
```
Comment (lorengordon) [inline salt/utils/templates.py]: Are we sure this works?

```
>>> repr('foo\n')
"'foo\\n'"
>>> repr('foo\n').endswith(('\n'))
False
>>> 'foo\n'
'foo\n'
>>> 'foo\n'.endswith(('\n'))
True
```

**Comment by lorengordon** [comment]
> I believe the python open command will replace line endings in a file with the system line endings.

I think it depends on how the file is opened, and whether using the [Universal Newlines][5] feature, which is implemented differently in [python2][4] and [python3][1]. If looking for a py2/3 compatible method for using universal newlines, I'd recommend the io.open library ([py2][2]/[py3][3]).

Not sure any of this is relevant to this particular pr though...

[1]: https://docs.python.org/3/library/functions.html#open
[2]: https://docs.python.org/2/library/io.html#io.open
[3]: https://docs.python.org/3/library/io.html#io.open
[4]: https://docs.python.org/2/library/functions.html#open
[5]: https://docs.python.org/3/glossary.html#term-universal-newlines

File: salt/utils/templates.py (hunk 1 above)
Comment (twangboy) [reply salt/utils/templates.py]: Yeah, it doesn't work... Let me see if I can figure out something else...

**Comment by lorengordon** [comment]
@rallytime I've looked at it, but I don't have an easy way to test the salt util functions... at least, not as easily as testing the builtin `repr`...

**Comment by twangboy** [comment]
@lorengordon You should be able to import the stringutils library:
```
>>> import salt.utils.stringutils
>>> import os
>>> salt.utils.stringutils.to_unicode('foo\n')
u'foo\n'
>>> salt.utils.stringutils.to_unicode('foo\n').endswith('\n')
True
>>> salt.utils.stringutils.to_unicode('foo\n').endswith(('\n', os.linesep))
True
```

**Comment by lorengordon** [comment]
@twangboy looks like a winner to me 😄

**Review by lorengordon — APPROVED** [review]

**Comment by damon-atkins** [comment]
FYI  info on `repr()` https://docs.djangoproject.com/en/1.11/topics/python3/#str-and-unicode-methods

**Review by rallytime — APPROVED** [review]

**Pull request merged**